test/conftest.py

```diff
@@ -1,9 +1,150 @@
+import json
+import subprocess
 import sys
+import time
+import uuid
 
 import pytest
+import requests
+from mock import ANY, patch
+from six.moves import queue
+from six.moves.urllib.parse import urlparse
 from types import ModuleType
 
 
+TOXIPROXY_HOST = "127.0.0.1"
+TOXIPROXY_PORT = 8474
+
+
+def pytest_configure(config):
+    config.addinivalue_line(
+        "markers",
+        "publish_retry: distinguish tests that should use retry in publishers"
+    )
+
+
+@pytest.yield_fixture
+def mock_producer():
+    with patch('nameko.amqp.producers') as patched:
+        with patched[ANY].acquire() as producer:
+            # normal behaviour is for no messages to be returned
+            producer.channel.returned_messages.get_nowait.side_effect = (
+                queue.Empty
+            )
+            yield producer
+
+
+@pytest.yield_fixture
+def mock_connection():
+    with patch('nameko.amqp.connections') as patched:
+        with patched[ANY].acquire() as connection:
+            yield connection
+
+
+@pytest.yield_fixture(scope='session')
+def toxiproxy_server():
+    # start a toxiproxy server
+    host = TOXIPROXY_HOST
+    port = TOXIPROXY_PORT
+    server = subprocess.Popen(
+        ['toxiproxy-server', '-port', str(port), '-host', host],
+        stdout=subprocess.PIPE
+    )
+    time.sleep(0.2)  # allow server to start
+    yield "{}:{}".format(host, port)
+    server.terminate()
+
+
+@pytest.yield_fixture
+def toxiproxy(toxiproxy_server, rabbit_config, free_port):
```

Comment:
hm. every user of the `free_port` fixture will be given the same port. which probably isn't what we want